### PR TITLE
feat(select): add compareWith property

### PR DESCRIPTION
**Pull request by zakton5**
#### Short description of what this resolves:
Using objects as values was only supported by comparing object equality (`===`) but not by content. This PR adds a `compareWith` property that accepts a property name (ie `id`) or a function of the shape `(o1, o2) => boolean` that is used to determine equality.

#### Changes proposed in this pull request:
- Add `compareWith` prop
- Move internal option comparison logic to its own function: `compareOptions`

**Ionic Version**: 4.0.0

**Fixes**: #14598

### CODE REVIEW

**Comment by remisture** [comment]
ETA for this?

**Comment by zakton5** [comment]
@remisture 🤷‍♂️ Whenever the Ionic team reviews it. Hard to say..

**Comment by brandyscarney** [comment]
Thank you for the PR!

I have this added to review for `4.1.0`. We were working out the process for how we would move forward with features so we haven't been adding them, but we are now! Although I'd almost consider this a fix as it's a regression. I have added myself and some other coworkers to review this.

My only request at first glance is could you please move the test out of the `preview/` folder - maybe we could change the basic test? We're planning to remove the `preview` folders that's why I ask. 🙂

**Comment by zakton5** [comment]
Will do @brandyscarney!

File: core/src/components/select/select.tsx [hunk 1]
```diff
@@ -468,33 +475,45 @@ function parseValue(value: any) {
...
+}
+
+function compareOptions(o1: any, o2: any, compareWith?: string | SelectCompareFn | null): boolean {
+  if (typeof compareWith === 'function') {
+    return compareWith(o1, o2);
+  } else if (typeof compareWith === 'string') {
+    return o1[compareWith] === o2[compareWith];
   } else {
-    return currentValue === optionValue;
+    return o1 === o2;
```
[the remark this answers is not in the document]
Comment (zakton5) [reply core/src/components/select/select.tsx]: Sure!

**Comment by brandyscarney** [comment]
Thank you @zakton5 !

File: core/src/components/select/select.tsx [hunk 2]
```diff
@@ -468,33 +475,45 @@ function parseValue(value: any) {
...
+  } else {
+    return compareOptions(currentValue, compareValue, compareWith);
+  }
+}
+
+function compareOptions(currentValue: any, compareValue: any, compareWith?: string | SelectCompareFn | null): boolean {
+  if (typeof compareWith === 'function') {
+    return compareWith(currentValue, compareValue);
+  } else if (typeof compareWith === 'string') {
+    return currentValue[compareWith] === compareValue[compareWith];
```
Comment (jaalzateolaya) [inline core/src/components/select/select.tsx:493]: This object comparison using a string is not null-safe. It doesn't check for the currentValue/compareValue to be set nor being an object.

**Pull request merged**